include/ignition/math/SpeedLimiter.hh

```diff
@@ -0,0 +1,147 @@
+#ifndef IGNITION_MATH_SYSTEMS_SPEEDLIMITER_HH_
+#define IGNITION_MATH_SYSTEMS_SPEEDLIMITER_HH_
+
+#include <chrono>
+#include <memory>
+#include <ignition/math/config.hh>
+#include "ignition/math/Helpers.hh"
+
+namespace ignition
+{
+namespace math
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_MATH_VERSION_NAMESPACE {
+  // Forward declaration.
+  class SpeedLimiterPrivate;
+
+  /// \brief Class to limit velocity, acceleration and jerk.
+  /// \ref https://github.com/ros-controls/ros_controllers/tree/melodic-devel/diff_drive_controller
+  class IGNITION_MATH_VISIBLE SpeedLimiter
+  {
+    /// \brief Constructor.
+    /// \param [in] _hasVelocityLimits     If true, applies velocity limits.
+    /// \param [in] _hasAccelerationLimits If true, applies acceleration limits.
+    /// \param [in] _hasJerkLimits         If true, applies jerk limits.
+    /// \param [in] _minVelocity Minimum velocity [m/s], usually <= 0. Defaults
+    ///     to negative infinity.
+    /// \param [in] _maxVelocity Maximum velocity [m/s], usually >= 0. Defaults
+    ///     to positive infinity.
+    /// \param [in] _minAcceleration Minimum acceleration [m/s^2], usually <= 0.
+    ///     Defaults to negative infinity.
+    /// \param [in] _maxAcceleration Maximum acceleration [m/s^2], usually >= 0.
+    ///     Defaults to positive infinity.
+    /// \param [in] _minJerk Minimum jerk [m/s^3], usually <= 0. Defaults to
+    ///     negative infinity.
+    /// \param [in] _maxJerk Maximum jerk [m/s^3], usually >= 0. Defaults to
+    ///     positive infinity.
+    public: SpeedLimiter(
+        bool   _hasVelocityLimits     = false,
+        bool   _hasAccelerationLimits = false,
+        bool   _hasJerkLimits         = false,
+        double _minVelocity           = -INF_D,
+        double _maxVelocity           = INF_D,
+        double _minAcceleration       = -INF_D,
+        double _maxAcceleration       = INF_D,
+        double _minJerk               = -INF_D,
+        double _maxJerk               = INF_D);
+
+    /// \brief Destructor.
+    public: ~SpeedLimiter();
+
+    /// \brief Limit velocity, acceleration and jerk, as enabled during
+    /// construction.
+    /// \param [in, out] _v Velocity to limit [m/s].
+    /// \param [in] _v0 Previous velocity to v  [m/s].
+    /// \param [in] _v1 Previous velocity to v0 [m/s].
+    /// \param [in] _dt Time step.
+    /// \return Limiting factor, which is (out_v / in_v).
+    public: double Limit(double &_v,
+                         double _v0,
+                         double _v1,
+                         std::chrono::steady_clock::duration _dt) const;
+
+    /// \brief Limit the velocity.
+    /// \param [in, out] _v Velocity to limit [m/s].
+    /// \return Limiting factor, which is (out_v / in_v).
+    public: double LimitVelocity(double &_v) const;
+
+    /// \brief Limit the acceleration.
+    /// \param [in, out] _v  Velocity [m/s].
+    /// \param [in] _v0 Previous velocity [m/s].
```

Comment:
how does `_vPrevious` sound?

Reply:
I updated all the variables to match the naming I was using on the tests.